### ISSUE TITLE
High term number on out-of-date group prevents it from rejoining Raft

### ISSUE THREAD

**Issue by tschottdorf** [cockroachdb/cockroach#1131]
found by @MBerhault:

- three node cluster
- kill one node, propose some commands on the rest, restart node (without network access though)
- the node will call elections repeatedly, increasing its term rapidly
- restore the network
- the node will be separated from the cluster:
  - it ignores incoming MsgApp (because others have lower term)
  - it cannot be elected leader (it keeps trying, but its log index is too low, i.e. it's missing data)

I haven't seen this discussed in the Raft paper. The right way out seems for the node to acknowledge an incoming MsgApp with a lower term (but a higher log index) as proof of a legitimate leader being out there, and working towards catching up its log. This however means that the term on the node will decrease. If before lowering the term we check that all the discarded terms are without progress, that may not have unintended side effects, but we need to think about it carefully - one Raft invariant is that the term number always increases and the above suggestion would amount to changing that invariant to read that the numbers of terms with a leader always increases.

It also appears that the Raft term wasn't persisted right away (i.e. when killing the node again and restarting, it loses all the "wasted" terms, which seems worth mentioning).

**Comment by tschottdorf** [comment]
via @bdarnell: The current leader should step down, increase its term and eventually an up-to-date replica would be elected and progress continue. We'll have to check why that didn't work. The correct fix are two-phase elections to avoid runaway term numbers.

**Comment by bdarnell** [comment]
This is discussed in section 9.6 of the thesis (https://github.com/ongardie/dissertation/blob/9cb493a23c15da7cfd0f6b375d16794b8bc568a5/leaderelection/prevote.tex)

**Comment by tschottdorf** [comment]
We need to fix this somehow, it's making cluster recovery very unstable. I can reliably kill two nodes, restart them and end up in a situation in which the only node which can be elected leader (by virtue of having the most up-to-date log) ends up with a term that forever trails the other two nodes' terms (all of which increase because no leader ever gets elected).

**Comment by tschottdorf** [comment]
Looks like there's something else at work here, the first node doesn't seem to be able to talk to nodes two and three after they restart. Likely caused by the fact that the client never manages to connect.

**Comment by bdarnell** [comment]
Is this higher priority than the panics (#1242, #1191, #1187) which are currently next in my queue?

**Comment by tschottdorf** [comment]
Not currently.

**Comment by bdarnell** [comment]
Panics are fixed so I'm taking a look at this next.

**Comment by bdarnell** [comment]
@tschottdorf What exactly is the procedure to reproduce this? I haven't been able to reproduce any persistent election failures, and I have seen the case where the two live nodes increase their term numbers to match the rejoining node's. 

I have, however, seen some cases where gossip seems to take longer than expected to get reconnected, which would cause the nodes to be unable to communicate. I think this is probably a gossip/transport problem instead of being related to the term numbers.

**Comment by tschottdorf** [comment]
let me try again @bdarnell, I think the transport problems were having a large impact.

**Comment by bdarnell** [comment]
Closing this since it hasn't been seen in a long time.
